src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -245,12 +246,103 @@ private BoundExpression ApplyConvertedTypes(BoundExpression expr, TupleBinaryOpe
+        /// In a tuple literal, each element can have an explicit name, an inferred name or no name.
+        /// In an expression of tuple type, each element can have a name or not.
+        /// </summary>
+        private static void ReportNamesMismatchesIfAny(BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            bool leftIsTupleLiteral = left.Kind == BoundKind.TupleLiteral;
+            bool rightIsTupleLiteral = right.Kind == BoundKind.TupleLiteral;
+
+            if (!leftIsTupleLiteral && !rightIsTupleLiteral)
+            {
+                return;
+            }
+
+            var (leftNames, leftInferred, leftLocations) = getDetails(left);
+            bool leftNoInferredNames = leftInferred.IsDefault;
+            bool leftNoNames = leftNames.IsDefault;
+
+            var (rightNames, rightInferred, rightLocations) = getDetails(right);
+            bool rightNoInferredNames = rightInferred.IsDefault;
+            bool rightNoNames = rightNames.IsDefault;
+
+            if (leftNoNames && rightNoNames)
+            {
+                return;
+            }
+
+            Debug.Assert(leftNoNames || rightNoNames || leftNames.Length == rightNames.Length);
+
+            int length = leftNoNames ? rightNames.Length : leftNames.Length;
+            for (int i = 0; i < length; i++)
+            {
+                string leftName = leftNoNames ? null : leftNames[i];
+                bool leftWasInferred = leftNoInferredNames ? false : leftInferred[i];
+
+                string rightName = rightNoNames ? null : rightNames[i];
+                bool rightWasInferred = rightNoInferredNames ? false : rightInferred[i];
+
+                bool different = string.CompareOrdinal(rightName, leftName) != 0;
+                if (!different)
+                {
+                    continue;
+                }
+
+                bool leftComplaint = leftIsTupleLiteral && leftName != null && !leftWasInferred;
+                bool rightComplaint = rightIsTupleLiteral && rightName != null && !rightWasInferred;
+
+                if (!leftComplaint && !rightComplaint)
+                {
+                    // No complaints, let's move on
+                    continue;
+                }
+
+                // When in doubt, we'll complain on the right side if it's a literal
+                bool useRight = (leftComplaint && rightComplaint) ? rightIsTupleLiteral : rightComplaint;
+                BoundExpression location = useRight ? rightLocations[i] : leftLocations[i];
+                string complaintName = useRight ? rightName : leftName;
+
+                diagnostics.Add(ErrorCode.WRN_TupleBinopLiteralNameMismatch, location.Syntax.Parent.Location, complaintName);
+            }
+
+            (ImmutableArray<string> NamesOpt, ImmutableArray<bool> InferredOpt, ImmutableArray<BoundExpression> LocationsOpt) getDetails(BoundExpression expr)
```

Comment:
>LocationsOpt [](start = 112, length = 12)

The name is misleading, these are not locations.

Reply:
It doesn't look like there is much benefit in having this field. I believe the code will be much clearer if we figure out location from tuple literal when we need it. Otherwise, we need to make sure consumers handle ```default``` value properly etc.

---
In reply to: [173941852](https://github.com/dotnet/roslyn/pull/25396#discussion_r173941852) [](ancestors = 173941852)